File: queue/prog.swift
import Foundation

class Stack<Element> {
  var items = [Element]()

  func push(_ el: Element) {
    items.append(el)
  }

  func pop() -> Element {
    return items.removeLast()
  }

  func tail() -> Element {
    return items.last!
  }

  func length() -> Int {
    return items.count
  }

  func empty() -> Bool {
    return length() == 0
  }
}

class Queue<Element> {
  var s1 = Stack<Element>()
  var s2 = Stack<Element>()

  func enqueue(_ el: Element) {
    s1.push(el)
  }

  func dequeue() {
    shift()
    _ = s2.pop()
  }

  func head() -> Element {
    shift()
    return s2.tail()
  }

  private

  func shift() {
    if s2.empty() {
       for _ in 0..<s1.length() {
         s2.push(s1.pop())
       }
    }
  }
}

let q = Int(readLine()!)!
var queue = Queue<Int>()
for i in 0..<q {
  let line = readLine()!.components(separatedBy: " ").map{ Int($0)! }
  var type = line[0]
  switch type {
  case 1:
    queue.enqueue(line[1])
  case 2:
    queue.dequeue()
  case 3:
    print("\(queue.head())")
  default:
    print("Unsupported Type")
  }
}

File: queue/prog.rb
#!/bin/ruby
q = gets.to_i

class Stack
  def initialize
    @items = []
  end

  def push(el)
    @items << el
  end

  def pop
    @items.pop
  end

  def tail
    @items.last
  end

  def length
    @items.length
  end

  def empty?
    length == 0
  end
end

class Queue
  def initialize
    @s1, @s2 = Stack.new, Stack.new
  end

  def enqueue(el)
    @s1.push el
  end

  def dequeue
    shift
    @s2.pop
  end

  def head
    shift
    @s2.tail
  end

  private

  def shift
    (@s1.length).times { @s2.push @s1.pop } if @s2.empty?
  end

end

queue = Queue.new
q.times do
  type, value = gets.split(" ").map(&:to_i)
  case type
  when 1
    queue.enqueue value
  when 2
    queue.dequeue
  when 3
    puts queue.head
  end
end
